src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java

```diff
@@ -143,7 +145,7 @@ public String toString() {
 
   @Override
   public BsonDocument clone() {
-    return getUnwrapped().clone();
```

Comment:
This is the source of the bug - there is no guarantee that unwrapped values are valid.  So create a new instance instead and unwrapping can happen lazily

Reply:
Sorry, one more question about this.  I'm not sure how we are using this method in the context of the project, but the javadoc for this overridden method says:

>  By convention, the object returned by this method should be independent of this object (which is being cloned).  To achieve this independence, it may be necessary to modify one or more fields of the object returned by {@code super.clone} before returning it.  Typically, this means copying any mutable objects that comprise the internal "deep structure" of the object being cloned and replacing the references to these objects with references to the copies.  If a class contains only primitive fields or references to immutable objects, then it is usually the case that no fields in the object returned by {@code super.clone} need to be modified.

This new implementation doesn't follow that convention.  Maybe it's ok because in the context of Kafka connector, all the fields of this class are effectively, if not actually, immutable?

Reply:
That's correct the mutability comes from the unwrapping which generates the mutable BsonDocument - that field isn't copied.

The reason this is required is the LazyBsonDocument should only unwrap if its actually required - clone doesn't require it.